lib/spack/llnl/util/filesystem.py

```diff
@@ -1823,59 +1823,50 @@ def find(root, files, recursive=True):
     if isinstance(files, str):
         files = [files]
 
-    if recursive:
-        tty.debug(f"Find (recursive): {root} {str(files)}")
-        result = _find_recursive(root, files)
-    else:
-        tty.debug(f"Find (not recursive): {root} {str(files)}")
-        result = _find_non_recursive(root, files)
+    if not recursive:
+        max_depth = 0
+
+    tty.debug(f"Find (max depth = {max_depth}): {root} {str(files)}")
+    result = find_max_depth(root, files, max_depth)
 
     tty.debug(f"Find complete: {root} {str(files)}")
     return result
 
 
-@system_path_filter
-def _find_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict. The idea is that
-    # we want to poke the filesystem as little as possible, but still maintain
-    # stability in the order of the answer. Thus we are recording each library
-    # found in a key, and reconstructing the stable order later.
-    found_files = collections.defaultdict(list)
-
-    # Make the path absolute to have os.walk also return an absolute path
-    root = os.path.abspath(root)
-    for path, _, list_files in os.walk(root):
-        for search_file in search_files:
-            matches = glob.glob(os.path.join(path, search_file))
-            matches = [os.path.join(path, x) for x in matches]
-            found_files[search_file].extend(matches)
+@system_path_filter(arg_slice=slice(1))
+def find_max_depth(root, globs, max_depth=None):
+    """Given a set of non-recursive glob file patterns, finds all
+    files matching those patterns up to a maximum specified depth.
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    Does not search in directories below the specified depth.
 
-    return answer
+    If ``globs`` is a list, files matching earlier entries are placed
+    in the return value before files matching later entries.
+    """
+    if not os.path.exists(root):
+        return []
 
+    root = root
+    if isinstance(globs, str):
+        globs = [globs]
+    regexes = [re.compile(fnmatch.translate(x)) for x in globs]
```

Comment:
Use the `"|".join(...)` trick to get a single regex?

Reply:
I had to undo the glob-to-regex translation because it turns out one of the unit tests, `test_pkg_attributes`, depends on the multi-dir behavior of globs (specifically on account of the `attributes-foo` `bar_headers` package). AFAIK nothing in the actual package repo depends on this, so I thought I'd try to take a step towards simplifying the interface, but missed this part.

I think we could go back to regexes (and change the test), in which case I think they could be combined into a single regex (although I'd need to talk some time to consider whether `|`-joining the patterns could produce something unexpected).

Reply:
Just read this, I think that's a bug in the test.

Reply:
Also note that you need `normcase` if you pursue this. See `class FindFirstFile`.

Reply:
I have restored the regex translation, and I have added a TODO for myself to verify that no packages depend on multi-dir search patterns (I think a full CI rebuild will be useful once we reach agreement on intended logic).

Reply:
Also thanks for the mention about `normcase`, although I don't understand why `FindFirstFile` uses it conditionally:

```
        if os.path is posixpath:
            self.match = regex.match
        else:
            self.match = lambda p: regex.match(os.path.normcase(p))
```

since I think `normcase` itself will already change its behavior based on the case-sensitivity of the FS.

Reply:
I think it was just to avoid two redundant function calls

Reply:
(This appears to be resolved to me, let me know if you disagree)